File: apps/server/src/sessionTitle.ts
import { readdir } from "node:fs/promises";
import { createReadStream } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import { createInterface } from "node:readline";
import type { ProviderId } from "./protocol.ts";

/**
 * Derives a human-readable "what is this session doing" title from the CLI's own
 * transcript files — the same data the CLI shows in its own session list.
 *
 *   - Claude writes an `ai-title` entry (a model-generated summary) into
 *     `~/.claude/projects/<encoded-cwd>/<cliSessionId>.jsonl`. We take the latest.
 *   - Codex has no generated title, so we fall back to the first `user_message`
 *     payload in its rollout file (the user's opening prompt).
 *
 * Returns null when nothing is available yet (e.g. before the first turn), in
 * which case the caller keeps the existing placeholder title.
 */

const CLAUDE_PROJECTS = join(homedir(), ".claude", "projects");
const CODEX_SESSIONS = join(homedir(), ".codex", "sessions");

const MAX_TITLE_LEN = 80;

/** Override the transcript roots (used by tests to point at fixtures). */
export interface TitleRoots {
  claudeProjects?: string;
  codexSessions?: string;
}

/**
 * Resolving a transcript file means scanning a whole directory tree, which is
 * wasteful to repeat on every poll. Cache the resolved path per CLI session id
 * once we've found it so later polls read just that one file.
 */
const fileCache = new Map<string, string>();

/** Collapse whitespace and trim/truncate a raw prompt or summary into a title. */
export function tidy(raw: string): string | null {
  const text = raw.replace(/\s+/g, " ").trim();
  if (!text) return null;
  return text.length > MAX_TITLE_LEN ? `${text.slice(0, MAX_TITLE_LEN - 1)}…` : text;
}

/** Find a `.jsonl` file by basename anywhere under `root`. */
async function findByBasename(root: string, basename: string): Promise<string | null> {
  let entries: string[];
  try {
    entries = await readdir(root, { recursive: true });
  } catch {
    return null;
  }
  const match = entries.find((e) => e.endsWith(basename));
  return match ? join(root, match) : null;
}

/** Stream JSONL lines, calling `onRecord`; stop early when it returns false. */
async function forEachRecord(
  file: string,
  onRecord: (rec: Record<string, unknown>) => boolean | void,
): Promise<void> {
  const stream = createReadStream(file, { encoding: "utf8" });
  const rl = createInterface({ input: stream, crlfDelay: Infinity });
  try {
    for await (const line of rl) {
      if (!line.trim()) continue;
      let rec: Record<string, unknown>;
      try {
        rec = JSON.parse(line) as Record<string, unknown>;
      } catch {
        continue;
      }
      if (onRecord(rec) === false) return;
    }
  } finally {
    rl.close();
    stream.destroy();
  }
}

/** Latest `ai-title` (model-generated summary) from a Claude transcript. */
export async function deriveClaudeTitle(
  cliSessionId: string,
  root: string = CLAUDE_PROJECTS,
): Promise<string | null> {
  let file = fileCache.get(cliSessionId);
  if (!file) {
    const found = await findByBasename(root, `${cliSessionId}.jsonl`);
    if (!found) return null;
    fileCache.set(cliSessionId, found);
    file = found;
  }
  let title: string | null = null;
  await forEachRecord(file, (rec) => {
    if (rec.type === "ai-title" && typeof rec.aiTitle === "string") {
      title = rec.aiTitle; // keep scanning — take the most recent one
    }
  });
  return title ? tidy(title) : null;
}

/** First user prompt from a Codex rollout, located by its session_meta id. */
export async function deriveCodexTitle(
  cliSessionId: string,
  root: string = CODEX_SESSIONS,
): Promise<string | null> {
  const cached = fileCache.get(cliSessionId);
  const files = cached ? [cached] : await codexRolloutFiles(root);

  for (const full of files) {
    let isMatch = false;
    let prompt: string | null = null;
    await forEachRecord(full, (rec) => {
      if (rec.type === "session_meta") {
        const payload = rec.payload as { id?: string } | undefined;
        if (payload?.id !== cliSessionId) return false; // wrong file — bail
        isMatch = true;
        return;
      }
      const payload = rec.payload as { type?: string; message?: string } | undefined;
      if (isMatch && payload?.type === "user_message" && typeof payload.message === "string") {
        prompt = payload.message;
        return false; // first user message is enough
      }
    });
    if (isMatch) {
      fileCache.set(cliSessionId, full);
      return prompt ? tidy(prompt) : null; // matched; null if no prompt yet
    }
  }
  return null;
}

/** Absolute paths of every Codex rollout file, newest scan each call. */
async function codexRolloutFiles(root: string): Promise<string[]> {
  let entries: string[];
  try {
    entries = await readdir(root, { recursive: true });
  } catch {
    return [];
  }
  return entries
    .filter((e) => e.endsWith(".jsonl") && e.includes("rollout-"))
    .map((e) => join(root, e));
}

export function deriveSessionTitle(
  provider: ProviderId,
  cliSessionId: string,
  roots: TitleRoots = {},
): Promise<string | null> {
  return provider === "claude"
    ? deriveClaudeTitle(cliSessionId, roots.claudeProjects)
    : deriveCodexTitle(cliSessionId, roots.codexSessions);
}

File: apps/native/Sources/JuancodeServices/SessionTitle.swift
import Foundation
import JuancodeCore

/// Derives a human-readable "what is this session doing" title from the CLI's own
/// transcript files — the same data the CLI shows in its own session list.
///
///   - Claude writes an `ai-title` entry (a model-generated summary) into
///     `~/.claude/projects/<encoded-cwd>/<cliSessionId>.jsonl`. We take the latest.
///   - Codex has no generated title, so we fall back to the first `user_message`
///     payload in its rollout file (the user's opening prompt).
///
/// Returns nil when nothing is available yet (e.g. before the first turn), in
/// which case the caller keeps the existing placeholder title.

public let CLAUDE_PROJECTS: String =
    (FileManager.default.homeDirectoryForCurrentUser.path as NSString)
    .appendingPathComponent(".claude/projects")
public let CODEX_SESSIONS: String =
    (FileManager.default.homeDirectoryForCurrentUser.path as NSString)
    .appendingPathComponent(".codex/sessions")

private let MAX_TITLE_LEN = 80

/// Override the transcript roots (used by tests to point at fixtures).
public struct TitleRoots {
    public var claudeProjects: String?
    public var codexSessions: String?
    public init(claudeProjects: String? = nil, codexSessions: String? = nil) {
        self.claudeProjects = claudeProjects
        self.codexSessions = codexSessions
    }
}

/// Resolving a transcript file means scanning a whole directory tree, which is
/// wasteful to repeat on every poll. Cache the resolved path per CLI session id
/// once we've found it so later polls read just that one file.
///
/// (Mirrors the module-level `Map` in the TS source; guarded by a lock since the
/// Swift module is reachable from multiple threads.)
private final class FileCache: @unchecked Sendable {
    private var map: [String: String] = [:]
    private let lock = NSLock()
    func get(_ key: String) -> String? {
        lock.lock(); defer { lock.unlock() }
        return map[key]
    }
    func set(_ key: String, _ value: String) {
        lock.lock(); defer { lock.unlock() }
        map[key] = value
    }
}
private let fileCache = FileCache()

/// Collapse whitespace and trim/truncate a raw prompt or summary into a title.
public func tidy(_ raw: String) -> String? {
    // Replace each run of whitespace with a single space, then trim. Mirrors the
    // JS `raw.replace(/\s+/g, " ").trim()`.
    let collapsed = raw
        .components(separatedBy: .whitespacesAndNewlines)
        .filter { !$0.isEmpty }
        .joined(separator: " ")
    let text = collapsed.trimmingCharacters(in: .whitespacesAndNewlines)
    if text.isEmpty { return nil }
    if text.count > MAX_TITLE_LEN {
        // `text.slice(0, MAX_TITLE_LEN - 1)` keeps 79 chars; appending the ellipsis
        // yields exactly MAX_TITLE_LEN characters.
        let prefix = String(text.prefix(MAX_TITLE_LEN - 1))
        return "\(prefix)…"
    }
    return text
}

/// Find a `.jsonl` file by basename anywhere under `root`.
public func findByBasename(_ root: String, _ basename: String) async -> String? {
    let entries = recursiveDirEntries(root)
    guard let match = entries.first(where: { $0.hasSuffix(basename) }) else { return nil }
    return (root as NSString).appendingPathComponent(match)
}

/// List every entry (files and dirs) under `root` recursively, as paths relative
/// to `root` — the Swift equivalent of `readdir(root, { recursive: true })`.
/// Returns an empty list when `root` cannot be read (mirrors the TS try/catch).
private func recursiveDirEntries(_ root: String) -> [String] {
    let fm = FileManager.default
    var isDir: ObjCBool = false
    guard fm.fileExists(atPath: root, isDirectory: &isDir), isDir.boolValue else { return [] }
    let rootURL = URL(fileURLWithPath: root, isDirectory: true)
    guard let enumerator = fm.enumerator(
        at: rootURL,
        includingPropertiesForKeys: nil,
        options: []
    ) else { return [] }
    var out: [String] = []
    let prefix = rootURL.standardizedFileURL.path
    for case let url as URL in enumerator {
        // Relative path from root, matching Node's recursive readdir output.
        let full = url.standardizedFileURL.path
        if full.hasPrefix(prefix + "/") {
            out.append(String(full.dropFirst(prefix.count + 1)))
        } else {
            out.append(url.lastPathComponent)
        }
    }
    return out
}

/// Stream JSONL lines, calling `onRecord`; stop early when it returns false.
public func forEachRecord(
    _ file: String,
    _ onRecord: ([String: Any]) -> Bool?
) async {
    // Read the whole file and split on newlines. We tolerate malformed lines
    // exactly as the TS does (skip lines that don't parse / are blank).
    guard let data = FileManager.default.contents(atPath: file),
          let contents = String(data: data, encoding: .utf8) else { return }
    // `crlfDelay: Infinity` in Node coalesces \r\n; splitting on \n and trimming
    // each line of trailing \r reproduces that.
    let lines = contents.split(separator: "\n", omittingEmptySubsequences: false)
    for rawLine in lines {
        let line = rawLine.hasSuffix("\r") ? String(rawLine.dropLast()) : String(rawLine)
        if line.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { continue }
        guard let lineData = line.data(using: .utf8) else { continue }
        let parsed = try? JSONSerialization.jsonObject(with: lineData)
        guard let rec = parsed as? [String: Any] else { continue }
        if onRecord(rec) == false { return }
    }
}

/// Latest `ai-title` (model-generated summary) from a Claude transcript.
public func deriveClaudeTitle(
    _ cliSessionId: String,
    _ root: String = CLAUDE_PROJECTS
) async -> String? {
    var file = fileCache.get(cliSessionId)
    if file == nil {
        guard let found = await findByBasename(root, "\(cliSessionId).jsonl") else { return nil }
        fileCache.set(cliSessionId, found)
        file = found
    }
    var title: String? = nil
    await forEachRecord(file!) { rec in
        if rec["type"] as? String == "ai-title", let aiTitle = rec["aiTitle"] as? String {
            title = aiTitle  // keep scanning — take the most recent one
        }
        return nil
    }
    if let title { return tidy(title) }
    return nil
}

/// First user prompt from a Codex rollout, located by its session_meta id.
public func deriveCodexTitle(
    _ cliSessionId: String,
    _ root: String = CODEX_SESSIONS
) async -> String? {
    let cached = fileCache.get(cliSessionId)
    let files = cached != nil ? [cached!] : await codexRolloutFiles(root)

    for full in files {
        var isMatch = false
        var prompt: String? = nil
        await forEachRecord(full) { rec in
            if rec["type"] as? String == "session_meta" {
                let payload = rec["payload"] as? [String: Any]
                if (payload?["id"] as? String) != cliSessionId { return false }  // wrong file — bail
                isMatch = true
                return nil
            }
            let payload = rec["payload"] as? [String: Any]
            if isMatch,
               payload?["type"] as? String == "user_message",
               let message = payload?["message"] as? String {
                prompt = message
                return false  // first user message is enough
            }
            return nil
        }
        if isMatch {
            fileCache.set(cliSessionId, full)
            if let prompt { return tidy(prompt) }  // matched; nil if no prompt yet
            return nil
        }
    }
    return nil
}

/// Absolute paths of every Codex rollout file, newest scan each call.
public func codexRolloutFiles(_ root: String) async -> [String] {
    let entries = recursiveDirEntries(root)
    return entries
        .filter { $0.hasSuffix(".jsonl") && $0.contains("rollout-") }
        .map { (root as NSString).appendingPathComponent($0) }
}

public func deriveSessionTitle(
    _ provider: ProviderId,
    _ cliSessionId: String,
    _ roots: TitleRoots = TitleRoots()
) async -> String? {
    if provider == .claude {
        return await deriveClaudeTitle(cliSessionId, roots.claudeProjects ?? CLAUDE_PROJECTS)
    } else {
        return await deriveCodexTitle(cliSessionId, roots.codexSessions ?? CODEX_SESSIONS)
    }
}
